### PR TITLE
v3

### LINKED ISSUE #28: Option to show line numbers
Hey Andrew,

you've built an amazing plugin, thanks a lot for your effort!
Is there a way to show line numbers for code snippets yet?

### LINKED ISSUE #51: Make line highlighting work without custom CSS
From the README:

> You need to pick your own background color, and optionally a left border width and color, for the highlighted lines. This can be done by setting CSS variables:

This is a bad default UX. We should be able to pick a reasonable highlight color from the theme text color. Dark themes are usually highlighted with ~10% opacity white; what are light themes usually highlighted with? 10% black is better than no default, but maybe we can pick one of the darker non-black syntax colors.

### LINKED ISSUE #62: Combine syntax highlighting and diff highlighting
I think I may have run into a use case for a new line transformer. I'd like to combine syntax highlighting and diff highlighting. As an example, this

````md
```js-diff
- export const oldNoop = () => {}
+ export const newNoop = () => {}
```
````

should render as a combination of regular highlights

```js
export const oldNoop = () => {}
export const newNoop = () => {}
```

plus added lines in green and removed lines in red.

```diff
- export const oldNoop = () => {}
+ export const newNoop = () => {}
```

What's your take on this? Would a line transformer be the best way to accomplish this?

### LINKED ISSUE #96: v3 Tracking Issue
There’s now enough momentum behind a few breaking changes that it’s time to start putting together a v3 branch! Progress in the v3 branch is being published to npm under `gatsby-remark-vscode@alpha`. Here’s the plan:

## New features!

- [x] Line transformers can contribute gutter cells - #95
- [x] Option to show line numbers (enabled by gutter cells) - #28, #95
- [x] Diff highlighting courtesy of @janosh (enabled by gutter cells) - #62, #88 

## Private APIs becoming public!

These need docs, tests, and final API refinements.

- [ ] Line transformers
- [ ] GraphQL resolvers

## Old APIs becoming deprecated!

This needs to be documented in MIGRATING.md.

- [x] `CodeBlockData['codeFenceNode']` (succeeded by `node` in #89)

## Deprecated APIs being removed!

- [x] Old `colorTheme` option

## Improvements!

- [x] Make the default CSS variables better - #51, #106 

## Not a blocker for v3 but nice to have

- [ ] Better logging/tracing - #91 
- [ ] Allow highlighting to be disabled for some code blocks - #79 
- [ ] Docs site with examples - #36 
- [ ] CLI to show info about extensions, languages, and themes - #90

### PR DESCRIPTION
**Pull request by andrewbranch**
Closes #28,
closes #51,
closes #62,
aaand closes #96.

Didn’t get all the docs and extra fancy features in I was hoping for, but I don’t want to make @janosh wait any longer. Thanks again @janosh for contributing an awesome feature! It looks great in the README:

image

### CODE REVIEW

**Comment by janosh** [comment]
@andrewbranch That's very kind of you! Honestly no need to rush v3 on my account though. 😄

On the other hand, this way we have some nice v3 minor releases to look forward to. 🚀

**Comment by janosh** [comment]
Just tried out v3 on my blog. For some reason, all the empty lines are gone. Currently investigating.

|  **v3** | **v2**  |
|---|---|
| v3 |  v2 |

**Comment by janosh** [comment]
Ah, so the empty lines are there. They just have 0 height. If I toggle off

```css
.grvsc-code {
    display: table;
}
```

they revert back to their former height.

File: MIGRATING.md [hunk 1]
```diff
@@ -1,5 +1,17 @@
 # Migration Guide
 
+## v2 → v3
+
+### Line highlighting is visible without custom CSS.
+
+Previously, the default line highlight background color was `transparent`, which meant a color had to be set with a CSS variable by the user. The reason for this limitation was that there is no sensible default that works for both light themes and dark themes, as a light highlight would be invisible on white, and a dark highlight would be invisible on black. As of v3, the plugin the luminance of each theme’s background color and selects a translucent white or translucent black highlight color as a default.
```
Comment (janosh) [inline MIGRATING.md:7]: Missing verb here: "...the plugin the luminance of ..."

**Comment by andrewbranch** [comment]
Oops, good catches! I’ll get a patch out momentarily. (`display: table` is necessary for gutters)

**Pull request merged**